streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskSchedulerImpl.scala

```diff
@@ -0,0 +1,180 @@
+
+  /**
+   * Get the resource requirements for all unscheduled tasks.
+   */
+  def getResourceRequests(): Array[ResourceRequest]
+
+  /**
+   * This notify the scheduler that a resource slot on {workerId} and {executorId} is allocated, and
+   * expect a task to run in return.
+   * Task locality should be considered when deciding whether to offer a task on target {worker}
+   * and {executor}
+   *
+   * @param workerId which worker this resource is located.
+   * @param executorId which executorId this resource belongs to.
+   * @return
+   */
+  def resourceAllocated(workerId : Int, executorId: Int) : Option[TaskLaunchData]
+
+  /**
+   * This notify the scheduler that {executorId} is failed, and expect a set of
+   * ResourceRequest for all failed tasks on that executor.
+   *
+   * @param executorId executor that failed
+   * @return
+   */
+  def executorFailed(executorId: Int) : Array[ResourceRequest]
+}
+
+class TaskSchedulerImpl(appId : Int, config: Config)  extends TaskScheduler {
+  private val LOG: Logger = LogUtil.getLogger(getClass, app = appId)
+
+  private var tasks = Map.empty[TaskId, TaskLaunchData]
+
+  // maintain a list to tasks to be scheduled
+  private var tasksToBeScheduled = Map.empty[Locality, mutable.Queue[TaskLaunchData]]
+
+  // find the locality of the tasks
+  private val taskLocator = new TaskLocator(config)
+
+  private var executorToTaskIds = Map.empty[Int, Set[TaskId]]
+
+  def setTaskDAG(dag: DAG): Unit = {
+    dag.tasks.foreach { params =>
+      val (taskGroupId, taskDescription) = params
+      0.until(taskDescription.parallelism).map((taskIndex: Int) => {
+        val taskId = TaskId(taskGroupId, taskIndex)
+        val locality = taskLocator.locateTask(taskDescription)
+        val taskLaunchData = TaskLaunchData(taskId, taskDescription, dag.subGraph(taskGroupId))
+        tasks += (taskId -> taskLaunchData)
+        scheduleTaskLater(taskLaunchData, locality)
+      })
+    }
+    val nonLocalityQueue = tasksToBeScheduled.getOrElse(NonLocality, mutable.Queue.empty[TaskLaunchData])
+    //reorder the tasks to distributing fairly on workers
+    val taskQueue = nonLocalityQueue.sortBy(_.taskId.index)
+    tasksToBeScheduled += (NonLocality -> taskQueue)
+  }
+
+  def getResourceRequests(): Array[ResourceRequest] ={
+    fetchResourceRequests(fromOneWorker = false)
+  }
+
+  import Relaxation._
+  private def fetchResourceRequests(fromOneWorker: Boolean = false): Array[ResourceRequest] ={
+    var resourceRequests = Array.empty[ResourceRequest]
+    tasksToBeScheduled.foreach(taskWithLocality => {
+      val (locality, tasks) = taskWithLocality
+      LOG.debug(s"locality : $locality, task queue size : ${tasks.size}")
+
+      if (tasks.size > 0) {
+      locality match {
+        case locality: WorkerLocality =>
+          resourceRequests = resourceRequests :+
+            ResourceRequest(Resource(tasks.size), locality.workerId, relaxation = SPECIFICWORKER)
+        case _ if fromOneWorker =>
```

Comment:
Do we account for `NonLocality` here ? `NonLocality` is an object rather than `case object`.